src/util/init.rs

```diff
@@ -92,70 +84,64 @@ where
-            array_base_ptr: *mut Item,
-            initialized_count: usize,
+        struct DropGuard<'a, Item> {
+            arr: &'a mut [MaybeUninit<Item>],
+            initialized: usize,
         }
 
-        impl<Item> Drop for DropGuard<Item> {
+        impl<Item> Drop for DropGuard<'_, Item> {
             fn drop(&mut self) {
                 // ## Safety
                 //
                 // The contract of the struct guarantees that this is sound
                 unsafe {
-                    let slice =
-                        slice::from_raw_parts_mut(self.array_base_ptr, self.initialized_count);
+                    let inited: &mut [Item] = self.arr[..self.initialized].assume_init_mut();
 
-                    ptr::drop_in_place(slice);
+                    // drop initialized elements
+                    ptr::drop_in_place(inited);
                 }
             }
         }
 
-        // If the `init(i)?` call panics or fails, `panic_guard` is dropped,
-        // dropping `array[.. initialized_count]` => no memory leak!
+        // If the `f(&mut state)?` call panics or fails, `guard` is dropped,
+        // thus dropping `array[..initialized]` => no memory leak!
         //
         // ## Safety
         //
-        // By construction, array[.. initialized_count] only contains
+        // By construction, `array[..initialized]` only contains
         // init elements, thus there is no risk of dropping uninit data.
-        unsafe {
-            let mut array = Arr::uninit();
-            let array_base_ptr = array.as_mut_slice().as_mut_ptr() as *mut Arr::Item;
-            let mut panic_guard = DropGuard {
-                array_base_ptr,
-                initialized_count: 0,
-            };
-
-            let mut state = init;
-            for i in 0..Arr::SIZE {
-                // Invariant: `i` elements have already been initialized
-                panic_guard.initialized_count = i;
-                // If this panics or fails, `panic_guard` is dropped, thus
-                // dropping the elements in `array_base_ptr[.. i]`
-                let value = f(&mut state)?;
-                // We can't use
-                // `*array.index_mut(i) = MaybeUninit::new(value);`
-                // here because of miri (see https://github.com/WaffleLapkin/arraylib/issues/5)
-                array_base_ptr.add(i).write(value)
-            }
+        let mut guard = DropGuard {
+            arr: array.as_mut_slice(),
+            initialized: 0,
+        };
+
+        // We need `guard` to hold unique reference to the `array`,
+        // so we can't access `array` directly
+        for (i, hole) in guard.arr.iter_mut().enumerate() {
+            // Invariant: `i` elements have already been initialized
+            guard.initialized = i;
+            // If `f(&mut state)?` panics or fails, `guard` is dropped, thus
+            // dropping the elements in `array[..i]`
+            *hole = MaybeUninit::new(f(&mut state)?);
```

Comment:
This two lines should be put in opposite order, otherwise it is trivial to introduce UB. If provided `f` _always_ panics, the very first iteration will set `guard.initialized` into `1` and later during stack unwinding `DropGuard::drop` creates `self.arr[..1].assume_init_mut()`, i. e. a mutable reference to slice which contains an item which was not properly initialized.

Reply:
1. this behaviour is saved from previous version 
2. no, they shouldn't, `.enumerate()` starts from _zero_ (0), so if `f` always panics, the very first iteration will set `guard.initialized` into **0** and later during stack unwinding `DropGuard::drop` creates `self.arr[..0].assume_init_mut()` which is totally ok since this is an empty slice

Reply:
Indeed, I completely forgot abot `.enumerate()` starting from zero. Disregard my comment above.